src/details/ArborX_Ray.hpp

```diff
@@ -198,6 +199,216 @@ bool intersects(Ray const &ray, Box const &box)
   return intersection(ray, box, tmin, tmax) && (tmax >= 0.f);
 }
 
+// rotate to the x-axis
+KOKKOS_INLINE_FUNCTION Point rotate2D(Point const &point)
+{
+  Point point_star;
+  float r = std::sqrt(point[0] * point[0] + point[1] * point[1]);
+  float costheta = 0.0f;
+  float sintheta = 0.0f;
+  if (r != 0.0f)
+  {
+    costheta = std::fabs(point[0]) / r;
+    sintheta = std::fabs(point[1]) / r;
+  }
+  point_star[0] = point[0] * costheta + point[1] * sintheta;
+  point_star[1] = point[2];
+  point_star[2] = 0.0;
+  return point_star;
+}
+
+// modified from Bruno's PR #604
+// https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection#Given_two_points_on_each_line_segment
+KOKKOS_INLINE_FUNCTION bool rayEdgeIntersect(Point const &edge_vertex_1,
+                                             Point const &edge_vertex_2,
+                                             float &t)
+{
+  float x3 = edge_vertex_1[0];
+  float y3 = edge_vertex_1[1];
+  float x4 = edge_vertex_2[0];
+  float y4 = edge_vertex_2[1];
+
+  float y1 = KokkosExt::min(y3, y4);
+  float y2;
+  if (y1 >= 0.f)
+  {
+    y2 = KokkosExt::max(y3, y4);
+  }
+  else
+  {
+    y2 = KokkosExt::min(y3, y4);
+  }
+  y1 = 0.f;
+
+  // float det = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4);
+  float det = -(y1 - y2) * (x3 - x4);
+
+  auto const epsilon = 0.00001f;
+  //  the ray is parallel to the edge if det == 0.0
+  //  When the ray overlaps the edge (x3==x4==0.0), it also returns false,
+  //  and the intersection will be captured by the other two edges.
+  if (det == 0.f)
+  {
+    return false;
+  }
+  // t = ((x1 - x3) * (y3 - y4) - (y1 - y3) * (x3 - x4)) / det;
+  t = (-x3 * (y3 - y4) - (y1 - y3) * (x3 - x4)) / det * y2;
+
+  if (t >= 0.f)
+  {
+    // float u = ((x1 - x3) * (y1 - y2) - (y1 - y3) * (x1 - x2)) / det;
+    float u = -x3 * (y1 - y2) / det;
+    if (u >= 0.f - epsilon && u <= 1.f + epsilon)
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+// The algorithm is described in
+// Watertight Ray/Triangle Intersection
+// [1] Woop, S. et al. (2013),
+// Journal of Computer Graphics Techniques Vol. 2(1)
+// The major difference is that here we return the intersection points
+// when the ray and the triangle is coplanar.
+// In the paper, they just need the boolean return.
+KOKKOS_INLINE_FUNCTION
+bool intersection(Ray const &ray, Triangle const &triangle, float &tmin,
+                  float &tmax)
+{
+  auto dir = ray.direction();
+  // normalizing the direction vector by its largest component.
+  // if precalculated, kx, ky, kz need to be stored.
+  int kz = 0;
+  for (int i = 1; i < 3; i++)
+  {
+    float compmax = std::fabs(dir[i - 1]);
+    if (std::fabs(dir[i]) > compmax)
+    {
+      compmax = dir[i];
+      kz = i;
+    }
+  }
+  int kx = kz + 1;
+  if (kx == 3)
+    kx = 0;
+  int ky = kx + 1;
+  if (ky == 3)
+    ky = 0;
```

Comment:
```suggestion
  // normalizing the direction vector by its largest component.
  // if precalculated, kx, ky, kz need to be stored.
  int kz = 2;
  for (int i = 0; i < 2; ++i)
    if (std::fabs(dir[i]) > std::fabs(dir[kz]))
      kz = i;
 int kx = (kz + 1) % 3;
 int ky = (kz + 2) % 3;
```

Reply:
The suggestion failed one test. I put the original code for finding the index for the largest component into a function.